L1Trigger/L1TTrackMatch/plugins/L1TkEmParticleProducer.cc

```diff
@@ -0,0 +1,391 @@
+
+    //float z_leadingTrack = -999;
+    //float Pt_leadingTrack = -999;
+
+    /*
+	if (DeltaZConstrain) {
+	// first loop over the tracks to find the leading one in DR < DRmax
+	for (trackIter = L1TTTrackHandle->begin(); trackIter != L1TTTrackHandle->end(); ++trackIter) {
+	float Pt = trackIter->momentum().perp();
+	float Eta = trackIter->momentum().eta();
+	float Phi = trackIter->momentum().phi();
+	float z  = trackIter->POCA().z();
+	if (fabs(z) > ZMAX) continue;
+	if (Pt < PTMINTRA) continue;
+	float chi2 = trackIter->chi2();
+	if (chi2 > CHI2MAX) continue;
+	float dr = deltaR(Eta, eta, Phi,phi);
+	if (dr < DRmax) {
+	if (Pt > Pt_leadingTrack) {
+	Pt_leadingTrack = Pt;
+	z_leadingTrack = z;
+	}
+	}
+	} // end loop over the tracks
+	} // endif DeltaZConstrain
+      */
+
+    for (trackIter = L1TTTrackHandle->begin(); trackIter != L1TTTrackHandle->end(); ++trackIter) {
+      float Pt = trackIter->momentum().perp();
+      float Eta = trackIter->momentum().eta();
+      float Phi = trackIter->momentum().phi();
+      float z = trackIter->POCA().z();
+      if (fabs(z) > ZMAX)
+        continue;
+      if (Pt < PTMINTRA)
+        continue;
+      float chi2 = trackIter->chi2();
+      if (chi2 > CHI2MAX)
+        continue;
+
+      float dr = deltaR(Eta, eta, Phi, phi);
+      if (dr < DRmax && dr >= DRmin) {
+        //std::cout << " a track in the cone, z Pt = " << z << " " << Pt << std::endl;
+        sumPt += Pt;
+      }
+
+      if (zvtxL1tk > -999 && fabs(z - zvtxL1tk) >= DeltaZMax)
+        continue;  // Now, PV constrained trackSum:
+
+      dr = deltaR(Eta, etaPV, Phi, phi);  // recompute using the corrected eta
+
+      if (dr < DRmax && dr >= DRmin) {
+        sumPtPV += Pt;
+      }
+    }  // end loop over tracks
+
+    if (RelativeIsolation) {
+      if (et > 0) {
+        trkisol = sumPt / et;
+        trkisolPV = sumPtPV / et;
+      }       // relative isolation
+    } else {  // absolute isolation
+      trkisol = sumPt;
+      trkisolPV = sumPtPV;
+    }
+
+    float isolation = trkisol;
+    if (PrimaryVtxConstrain) {
+      isolation = trkisolPV;
+    }
+
+    const math::XYZTLorentzVector P4 = egIter->p4();
+    TkEm trkEm(P4, EGammaRef, trkisol, trkisolPV);
+
+    if (IsoCut <= 0) {
+      // write the L1TkEm particle to the collection,
+      // irrespective of its relative isolation
+      result->push_back(trkEm);
+    } else {
+      // the object is written to the collection only
+      // if it passes the isolation cut
+      if (isolation <= IsoCut)
+        result->push_back(trkEm);
+    }
+
+  }  // end loop over EGamma objects
+
+  iEvent.put(std::move(result), label);
+}
+
+// --------------------------------------------------------------------------------------
+
+float L1TkEmParticleProducer::CorrectedEta(float eta, float zv) {
+  // Correct the eta of the L1EG object once we know the zvertex
+
+  bool IsBarrel = (fabs(eta) < 1.479);
+  float REcal = 129.;
```

Comment:
can these be obtained from geometry? if not, constexpr

Reply:
done